Feature/foreign keys

JakeSchoermer opened the pull request:
Added support for foreign keys (available in Rails 4.2+) along with relevant shared example specs and documentation.

bradleybeddoes commented on lib/gumboot/shared_examples/foreign_keys.rb:
```diff
@@ -0,0 +1,51 @@
+    let(:permission_fk) do
+      conn.foreign_keys(:permissions).find do |fk|
+        fk.options[:column] == 'role_id'
+      end
```
The pattern on L5 - L7 is duplicated 5 times in this file, you should be able to abstract this to a single method that takes in `table_name` and `column`.

bradleybeddoes commented on lib/gumboot/shared_examples/foreign_keys.rb:
```diff
@@ -0,0 +1,51 @@
+RSpec.shared_examples 'Foreign Keys' do
+  conn = ActiveRecord::Base.connection
+  context 'Permissions Foreign Keys', if: conn.supports_foreign_keys? do
```
Can the `if` statement you have here, which is duplicated through the file be applied once at the `shared_examples` level by chance?

bradleybeddoes commented on lib/gumboot/shared_examples/foreign_keys.rb:
```diff
@@ -0,0 +1,51 @@
+    it 'should have a foreign key' do
+      expect(permission_fk).to_not be_nil
+    end
+  end
```
Seems like there should be additional specs to ensure the correct 'to' table? In this case for example `role_id` points to `role`. Is that doable?

JakeSchoermer commented:
Hi @bradleybeddoes, I've addressed the concerns you raised and the PR ready for you to have another look at.

bradleybeddoes commented on lib/gumboot/shared_examples/foreign_keys.rb:
```diff
@@ -0,0 +1,71 @@
+      end
+      it 'should have a foreign key that points to the correct table' do
+        expect(fk_exists_between('permissions', 'roles')).to be true
+      end
```
L6 - 11 is essentially duplicated in every context here.

Please pull this out into a shared example so the code appears exactly one, and re-use it.

bradleybeddoes commented on lib/gumboot/shared_examples/foreign_keys.rb:
```diff
@@ -0,0 +1,71 @@
+    end
+
+    def table_has_fk(table, foreign_key)
+      f = ActiveRecord::Base.connection.foreign_keys(table).find do |fk|
```
could this not just use `conn` which you've already defined?

bradleybeddoes commented on lib/gumboot/shared_examples/foreign_keys.rb:
```diff
@@ -0,0 +1,71 @@
+
+    def fk_exists_between(from_table, to_table)
+      f = ActiveRecord::Base.connection.foreign_keys(from_table).find do |fk|
+        fk[:to_table] == to_table
```
I would suggest re-checking column name here, so we're certain that it is the expected foreign key that points to `:to_table` and not something else that was introduced to the model.

bradleybeddoes commented on lib/gumboot/shared_examples/foreign_keys.rb:
```diff
@@ -0,0 +1,71 @@
+    end
+
+    def fk_exists_between(from_table, to_table)
+      f = ActiveRecord::Base.connection.foreign_keys(from_table).find do |fk|
```
The use of `f` in this method and the resultant `return` confuses the implementation.

This method can be improved, likely in partnership with rspec `.to be_truthy | be_falesey` which you should be using over `.to be true|false`

The same applies for `table_has_fk`.

bradleybeddoes commented on lib/gumboot/shared_examples/foreign_keys.rb:
```diff
@@ -0,0 +1,71 @@
+        it 'should have a foreign key that points to the correct table' do
+          expect(
+            fk_exists_between('api_subject_roles', 'api_subjects')
+          ).to be true
```
This would be cleaner as

``` ruby
expect(fk_exists_between('api_subject_roles', 'api_subjects'))
  .to be_truthy
```

JakeSchoermer commented:
@bradleybeddoes, I've responded to the second round of suggestions from you. It's much cleaner now.

I do think that the `table_has_fk` is now probably redundant, and can be removed given that the `fk_exists_between` method checks for both a foreign key and a both tables, but happy to take your advice on this,

bradleybeddoes commented:
@JakeSchoermer both specs are valid, they both tell a different story.

However this commit needs further refinement. What you've arrived at now completely breaks the documentation you've proposed for gumboot when working with client apps within this PR.

Please revisit these changes, the documentation as it is written right now should be valid for client apps, that is you need to look at changing your most recent code, not the documentation.

JakeSchoermer commented:
Ok @bradleybeddoes, I've re-done this to conform with the documentation whilst maintaining concision.

bradleybeddoes commented:
The below is the result of using your proposed changes in a client app. The fact that I don't have the foreign keys present is not the issue. This is seriously concerning given the importance of Gumboot to the AAF application stack.

``` ruby
$ rspec spec/models/gumboot_foreign_keys_spec.rb
I, [2016-04-22T13:48:37.102756 #44500]  INFO -- : Not reporting to Code Climate because ENV['CODECLIMATE_REPO_TOKEN'] is not set.
Run options: include {:focus=>true}

All examples were filtered out; ignoring {:focus=>true}

Randomized with seed 43695

Foreign Keys
  Permissions Foreign Keys
    should have a foreign key that points to the correct table (FAILED - 1)
    should have a foreign key (FAILED - 2)
  Permissions Foreign Keys
    should have a foreign key that points to the correct table (FAILED - 3)
    should have a foreign key (FAILED - 4)
  Permissions Foreign Keys
    should have a foreign key (FAILED - 5)
    should have a foreign key that points to the correct table (FAILED - 6)
  Permissions Foreign Keys
    should have a foreign key (FAILED - 7)
    should have a foreign key that points to the correct table (FAILED - 8)
  Permissions Foreign Keys
    should have a foreign key that points to the correct table (FAILED - 9)
    should have a foreign key (FAILED - 10)
```

Once again I would request that you reconsider your implementation and give further consideration to the usage of both `include_examples` and `let` as provided by RSpec.

bradleybeddoes commented:
Additionally specs provided in this PR will only work with ActiveRecord. This is our default ORM but we do have some usage of Sequel also.

The documentation makes no note of that so when adding it to a Sequel project the following occurs:

```
$ rspec spec/models/gumboot_foreign_keys_spec.rb

.../.rbenv/versions/2.2.3/lib/ruby/gems/2.2.0/bundler/gems/aaf-gumboot-046b6b3cf92f/lib/gumboot/shared_examples/foreign_keys.rb:2:in `conn': uninitialized constant ActiveRecord (NameError)
```

Please make an explict note in the documentation that this spec is only valid for users of ActiveRecord and that applications using Sequel should implement these specifications themselves.

JakeSchoermer commented:
@bradleybeddoes I've addressed the issues with the formatting of the output for my specs and the README docs. In your last comment you said:

> Once again I would request that you reconsider your implementation and give further consideration to the usage of both include_examples and let as provided by RSpec.

I am uncertain however, what you meant by include_example and let. Can you give me any more information?

bradleybeddoes commented:
I have written an example of how this should be implemented in terms of my comments above.

I don't guarantee this is 100% functional, you may need to tweak something, especially around L17.

https://gist.github.com/bradleybeddoes/d3a940adc5974cb177e97c59964ba68a

Not only is this much cleaner to read, it also outputs better. It also has the benefit of being extensible should some additional requirement come up that is only applicable to a subset of the tables.

JakeSchoermer commented:
Thanks for the help @bradleybeddoes. I've reimplemented my specs based on your suggestion with a few small changes and updated the relevant documentation. Happy for you to review again at your earliest convenience.

bradleybeddoes commented on lib/gumboot/shared_examples/foreign_keys.rb:
```diff
@@ -0,0 +1,65 @@
+      let(:from_table) { 'permissions' }
+      let(:to_table) { 'roles' }
+      let(:column) { 'role_id' }
+    end
```
This should have a context of `Roles` to match the other implementations

bradleybeddoes commented on lib/gumboot/shared_examples/foreign_keys.rb:
```diff
@@ -0,0 +1,65 @@
+        let(:column) { 'role_id' }
+      end
+    end
+    context 'API Subject' do
```
Should be plural

bradleybeddoes commented on lib/gumboot/shared_examples/foreign_keys.rb:
```diff
@@ -0,0 +1,65 @@
+      include_examples 'gumboot fk' do
+        let(:from_table) { 'subject_roles' }
+        let(:to_table) { 'roles' }
+        let(:column) { 'role_id' }
```
53 and 54 are duplicated at 60 and 61

JakeSchoermer commented:
Hi @bradleybeddoes. I've dealt with these small issues raised a couple of hours ago.
